File: Semestrul 4/MPP/lab06/cs/Repository/Impl/RefereeRepositoryImpl.cs
using System;
using System.Data;
using System.Data.SQLite;
using cs.Domain;
using cs.Repository.Abstracts;
using cs.Repository.Interfaces;

namespace cs.Repository.Impl
{
    public class RefereeRepositoryImpl : AbstractDBRepository<long, Referee>, IRefereeRepository
    {
        public RefereeRepositoryImpl() : base("Referees", "id")
        {
        }

        protected override string GetInsertParameters()
        {
            return "username, password, email, first_name, last_name";
        }

        protected override string GetUpdateParameters()
        {
            return "username=@username, password=@password, email=@email, first_name=@firstName, last_name=@lastName";
        }

        protected override int GetUpdateParameterCount()
        {
            return 5;
        }

        protected override Referee CreateEntity(IDataReader reader)
        {
            long id = Convert.ToInt64(reader["id"]);
            string username = reader["username"].ToString();
            string password = reader["password"].ToString();
            string email = reader["email"].ToString();
            string firstName = reader["first_name"].ToString();
            string lastName = reader["last_name"].ToString();

            Referee referee = new Referee(username, password, email, firstName, lastName);
            referee.SetId(id);
            return referee;
        }

        protected override void SetParameters(SQLiteCommand command, Referee entity)
        {
            command.Parameters.AddWithValue("@username", entity.GetUsername());
            command.Parameters.AddWithValue("@password", entity.GetPassword());
            command.Parameters.AddWithValue("@email", entity.GetEmail());
            command.Parameters.AddWithValue("@firstName", entity.GetFirstName());
            command.Parameters.AddWithValue("@lastName", entity.GetLastName());
        }

        public Referee? FindByUsername(string username)
        {
            string sql = $"SELECT * FROM {_tableName} WHERE username = @username";

            try
            {
                using var command = new SQLiteCommand(sql, _connection);
                command.Parameters.AddWithValue("@username", username);

                using var reader = command.ExecuteReader();
                if (reader.Read())
                {
                    return CreateEntity(reader);
                }
            }
            catch (Exception ex)
            {
                // Logger.LogError($"Error finding referee by username: {ex.Message}");
            }

            return null;
        }
    }
}

File: Semestrul 4/MPP/lab04/java/src/main/java/mpp/repository/impl/RefereeRepositoryImpl.java
package mpp.repository.impl;

import mpp.domain.Referee;
import mpp.repository.abstracts.AbstractDBRepository;
import mpp.repository.interfaces.RefereeRepository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Optional;

public class RefereeRepositoryImpl extends AbstractDBRepository<Long, Referee> implements RefereeRepository {
    public RefereeRepositoryImpl() {
        super("Referees", "id");
    }

    @Override
    protected String getInsertParameters() {
        return "username, password, email, first_name, last_name";
    }

    @Override
    protected String getUpdateParameters() {
        return "username=?, password=?, email=?, first_name=?, last_name=?";
    }

    @Override
    protected int getUpdateParameterCount() {
        return 5;
    }

    @Override
    protected Referee createEntity(ResultSet data) throws SQLException {
        Long id = data.getLong("id");
        String username = data.getString("username");
        String password = data.getString("password");
        String email = data.getString("email");
        String firstName = data.getString("first_name");
        String lastName = data.getString("last_name");

        Referee referee = new Referee(username, password, email, firstName, lastName);
        referee.setId(id);
        return referee;
    }

    @Override
    protected void saveEntity(PreparedStatement preparedStatement, Referee entity) throws SQLException {
        preparedStatement.setString(1, entity.getUsername());
        preparedStatement.setString(2, entity.getPassword());
        preparedStatement.setString(3, entity.getEmail());
        preparedStatement.setString(4, entity.getFirstName());
        preparedStatement.setString(5, entity.getLastName());
    }

    @Override
    public Optional<Referee> findByUsername(String username) {
        String sql = "SELECT * FROM " + tableName + " WHERE username = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, username);

            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    return Optional.of(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding referee by username: " + e.getMessage());
        }

        return Optional.empty();
    }
}